cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -0,0 +1,472 @@
+      throws OptimizationFailureException {
+    _mustHaveTopicLeadersPerBroker = Collections.unmodifiableSet(
+        Utils.getTopicNamesMatchedWithPattern(_balancingConstraint.topicsWithMinLeadersPerBrokerPattern(), clusterModel::topics));
+    if (_mustHaveTopicLeadersPerBroker.isEmpty()) {
+      return;
+    }
+    // Sanity checks
+    validateTopicsWithMinLeaderReplicaIsNotExcluded(optimizationOptions);
+    validateEnoughLeaderReplicaToDistribute(clusterModel, optimizationOptions);
+    validateBrokersAllowedReplicaMoveExist(clusterModel, optimizationOptions);
+    boolean onlyMoveImmigrantReplicas = optimizationOptions.onlyMoveImmigrantReplicas();
+    new SortedReplicasHelper().maybeAddSelectionFunc(ReplicaSortFunctionFactory.selectImmigrants(), onlyMoveImmigrantReplicas)
+                              .addSelectionFunc(ReplicaSortFunctionFactory.selectReplicasBasedOnIncludedTopics(_mustHaveTopicLeadersPerBroker))
+                              .maybeAddPriorityFunc(ReplicaSortFunctionFactory.prioritizeImmigrants(), !onlyMoveImmigrantReplicas)
+                              .trackSortedReplicasFor(_replicaSortName, clusterModel);
+  }
+
+  private int minTopicLeadersPerBroker() {
+    return _balancingConstraint.minTopicLeadersPerBroker();
+  }
+
+  private void validateTopicsWithMinLeaderReplicaIsNotExcluded(OptimizationOptions optimizationOptions)
+      throws OptimizationFailureException {
+    if (optimizationOptions.excludedTopics().isEmpty()) {
+      return;
+    }
+    Set<String> shouldNotBeExcludedTopics = new HashSet<>();
+    _mustHaveTopicLeadersPerBroker.forEach(topicName -> {
+      if (optimizationOptions.excludedTopics().contains(topicName)) {
+        shouldNotBeExcludedTopics.add(topicName);
+      }
+    });
+    if (!shouldNotBeExcludedTopics.isEmpty()) {
+      throw new OptimizationFailureException(String.format("Topics that must have a minimum number of leader replicas "
+          + "per alive broker that is not excluded for leadership moves should not be excluded. This error implies "
+          + "a config error. Topics should not be excluded=[%s] and it is defined as a regex by the config property: %s",
+                                                           String.join(", ", shouldNotBeExcludedTopics),
+                                                           AnalyzerConfig.TOPICS_WITH_MIN_LEADERS_PER_BROKER_CONFIG));
+    }
+  }
+
+  private void validateEnoughLeaderReplicaToDistribute(ClusterModel clusterModel,
+                                                       OptimizationOptions optimizationOptions) throws OptimizationFailureException {
+    Map<String, Integer> numLeadersByTopicNames = clusterModel.numLeadersPerTopic(_mustHaveTopicLeadersPerBroker);
+    Set<Broker> eligibleBrokersForLeadership = eligibleBrokersForLeadership(clusterModel, optimizationOptions);
+    int totalMinimumLeaderReplicaCount = eligibleBrokersForLeadership.size() * minTopicLeadersPerBroker();
```

Comment:
Nit (I will not list each case here, but this applies to any other function name, variable name, and comment in this class): Can we change `LeaderReplica` -> `Leader`?

Reply:
Fixed